internal/api/fetch_test_plan.go

```diff
@@ -0,0 +1,58 @@
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/buildkite/test-splitter/internal/plan"
+)
+
+type errorResponse struct {
+	Message string `json:"message"`
+}
+
+// FetchTestPlan fetchs a test plan from the server's cache.
+// If the test plan is found in the cache, it is returned, otherwise nil is returned.
+// Error is returned if there is a client side failure or request is invalid (400 - 403).
+// Other server errors are ignored and treated as cache miss.
+func (c client) FetchTestPlan(suiteSlug string, identifier string) (*plan.TestPlan, error) {
+	url := fmt.Sprintf("%s/v2/analytics/organizations/%s/suites/%s/test_plan?identifier=%s", c.ServerBaseUrl, c.OrganizationSlug, suiteSlug, identifier)
+
+	// send request
+	resp, err := c.httpClient.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("sending request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	switch {
+	case resp.StatusCode == http.StatusOK:
+		// happy path
+	case resp.StatusCode >= 400 && resp.StatusCode <= 403:
+		// treat 400-403 as an error because the request is invalid
+		responseBody, _ := io.ReadAll(resp.Body)
+		var errorResp errorResponse
+		json.Unmarshal(responseBody, &errorResp)
+		return nil, fmt.Errorf(errorResp.Message)
+	default:
+		// ignore other errors and treat them as cache miss
+		return nil, nil
+	}
```

Comment:
@wooly @niceking I'm ignoring server error, except for 400 ~ 403, so the client can proceed and attempt to make POST request. Do you think this make sense? Or should we implement a retry mechanism for this as well?

The reason behind handling 400 ~ 403 is that we want to terminate the program if there is bad request (miss config), unauthorized, or not permitted (token doesn't have required scope)

Reply:
I haven't fully thought through the implications of falling back to cache-miss yet, but it probably makes sense to document the individual status codes that we respond to in the splitter and what they mean. At least to have as a point of reference, even if the code groups them all together.

Reply:
Good suggestion. I have improve the readability and the test. Let me know what you think.